source/custom/visa_service.custom.cpp

```diff
@@ -0,0 +1,440 @@
+static std::shared_mutex s_resource_manager_lock;
+
+// Returns true if it's safe to use outputs of a method with the given status.
+inline bool status_ok(int32 status)
+{
+  return status >= VI_SUCCESS;
+}
+
+class DriverErrorException : public std::runtime_error
+{
+  private:
+    int status_ = 0;
+
+  public:
+    DriverErrorException(int status) : std::runtime_error(""), status_(status)
+    {
+    }
+    int status() const
+    {
+      return status_;
+    }
+};
+
+static ViSession GetResourceManagerSession(visa_grpc::VisaService::LibrarySharedPtr library)
+{
+    std::unique_lock<std::shared_mutex> lock(s_resource_manager_lock);
+    if (!s_defaultRM)
+    {
+        ViStatus status = library->OpenDefaultRM(&s_defaultRM);
+        if (!status_ok(status))
+        {
+            throw DriverErrorException(status);
+        }
+    }
+    return s_defaultRM;
+}
+
+::grpc::Status VisaService::ConvertApiErrorStatusForViSession(::grpc::ServerContextBase* context, int32_t status, ViSession vi)
+{
+  std::string description(nidevice_grpc::kMaxGrpcErrorDescriptionSize, '\0');
+  library_->StatusDesc(vi, status, &description[0]);
+  return nidevice_grpc::ApiErrorAndDescriptionToStatus(context, status, description);
+}
+
+::grpc::Status VisaService::ConvertApiErrorStatusForViObject(::grpc::ServerContextBase* context, int32_t status, ViObject vi)
+{
+  std::string description(nidevice_grpc::kMaxGrpcErrorDescriptionSize, '\0');
+  library_->StatusDesc(vi, status, &description[0]);
+  return nidevice_grpc::ApiErrorAndDescriptionToStatus(context, status, description);
+}
+
+//---------------------------------------------------------------------
+  //---------------------------------------------------------------------
+  ::grpc::Status VisaService::Close(::grpc::ServerContext* context, const CloseRequest* request, CloseResponse* response)
+  {
+    if (context->IsCancelled()) {
+      return ::grpc::Status::CANCELLED;
+    }
+    try {
+      ViStatus status;
+      if (request->object_handle().has_vi())
+      {
+        auto vi_grpc_session = request->object_handle().vi();
+        ViSession vi = session_repository_->access_session(vi_grpc_session.name());
+        session_repository_->remove_session(vi_grpc_session.name());
+        status = library_->Close(vi);
+        if (!status_ok(status)) {
+        return ConvertApiErrorStatusForViObject(context, status, vi);
+      }
+      }
+      else if (request->object_handle().has_object_handle())
+      {
+        auto object_handle = request->object_handle().object_handle();
+        status = library_->Close(object_handle);
+        if (!status_ok(status)) {
+        return ConvertApiErrorStatusForViObject(context, status, object_handle);
+        }
+      }
+      response->set_status(status);
+      return ::grpc::Status::OK;
+    }
+    catch (nidevice_grpc::NonDriverException& ex) {
+      return ex.GetStatus();
+    }
+    catch (const DriverErrorException& ex) {
```

Comment:
Nothing in this method throws this type (unlike the others that use `GetResourceManagerSession`). It's unclear to me whether it's better to be precise or make them all the same for the sake of safety and readability.

Reply:
Commit 5